### PR TITLE
[Prim][PIR]Flatten decomposite rule support dynamic shape

### PR DESCRIPTION
**Pull request by zeroRains**
### PR types

New features
### PR changes

Others
### Description

Flatten decomposite rule support dynamic shape

### CODE REVIEW

File: paddle/fluid/primitive/composite/composite.h [hunk 1]
```diff
@@ -736,31 +736,66 @@ std::tuple<Tensor, Tensor> flatten_decomp(const Tensor& x,
...
+    if (end_axis == start_axis) {
+      return std::make_tuple(backend::reshape<T>(x, x_shape), x_shape_tensor);
+    }
+    std::vector<Tensor> out_shape;
+    for (size_t i = 0; i < x_dim.size();) {
+      if (i >= static_cast<size_t>(start_axis) &&
+          i <= static_cast<size_t>(end_axis)) {
+        Tensor flat = get_slice<T>(x_shape, i);
+        i++;
+        while (i <= static_cast<size_t>(end_axis)) {
```
Comment (cyber-pioneer) [inline paddle/fluid/primitive/composite/composite.h]: end_axis考虑处理-1的情况

Comment (zeroRains) [reply paddle/fluid/primitive/composite/composite.h]: python端应该处理好了end_axis和start_axis为负数的情况
image

File: paddle/fluid/primitive/composite/composite.h [hunk 2]
```diff
@@ -736,31 +736,66 @@ std::tuple<Tensor, Tensor> flatten_decomp(const Tensor& x,
...
+    x_shape_tensor =
+        backend::full_with_tensor<T>(x_shape_tensor, 0.0, DataType::FLOAT32);
+    if (end_axis == start_axis) {
+      return std::make_tuple(backend::reshape<T>(x, x_shape), x_shape_tensor);
+    }
+    std::vector<Tensor> out_shape;
+    for (size_t i = 0; i < x_dim.size();) {
+      if (i >= static_cast<size_t>(start_axis) &&
+          i <= static_cast<size_t>(end_axis)) {
+        Tensor flat = get_slice<T>(x_shape, i);
```
Comment (cyber-pioneer) [inline paddle/fluid/primitive/composite/composite.h]: simplify like : slice<T>(x, {0}, {start_idx}, {end_idx}, {1}, {});

File: paddle/fluid/primitive/composite/composite.h [hunk 3]
```diff
@@ -736,31 +736,66 @@ std::tuple<Tensor, Tensor> flatten_decomp(const Tensor& x,
...
+        i++;
+        while (i <= static_cast<size_t>(end_axis)) {
+          flat = flat * get_slice<T>(x_shape, i);
```
Comment (cyber-pioneer) [inline paddle/fluid/primitive/composite/composite.h]: simplify like this: prod<T>(axis_dims, {0}, false, false);

File: paddle/fluid/primitive/composite/composite.h (hunk 1 above)
Comment (cyber-pioneer) [reply paddle/fluid/primitive/composite/composite.h]: c++端最好也处理一下，有些program的获取是不经过python的

File: paddle/fluid/primitive/composite/composite.h (hunk 2 above)
Comment (zeroRains) [reply paddle/fluid/primitive/composite/composite.h]: Done

File: paddle/fluid/primitive/composite/composite.h (hunk 3 above)
Comment (zeroRains) [reply paddle/fluid/primitive/composite/composite.h]: Done

**Review by cyber-pioneer — APPROVED** [review]

**Pull request merged**